### DISCUSSION TITLE
Network topology overview

### DISCUSSION THREAD

**Discussion by romain-fontugne** [InternetHealthReport/gsoc#36] [Ideas]
IHR monitors the inter-dependence of networks (e.g. the 
university of Tokyo relies on Japan's academic ISP) and displays these
dependencies with very simple line charts that highlights changes over time.
Representing this data as a graph is more intuitive for network operators.
So the goal of this project is to show the connectivity of a network (aka AS) in
a simple graph.
Using data from the Internet Yellow Pages we can depict a very detailed view of
an AS, by finding all prefixes originated by the AS, fetching AS dependencies for
all these prefixes, grouping these prefixes based on their dependencies, and 
finally displaying an AS graph that shows how these prefixes are connected to the
Internet. Additional information for each AS/prefix can be displayed by querying
IYP.
Related to this project, we have an offline tool that shows network dependencies 
for a country, this is also something we'd like to implement on IHR website
(https://github.com/InternetHealthReport/country-as-hegemony-viz).

Expected results:
- Create a VueJS component that show the dependency graph
- Integrate this component in IHR network report
- (optional) Create similar views for a country or a single prefix

**Comment by Animesh239** [comment]
`Additional information for each AS/prefix can be displayed by querying
IYP.`
Querying from IYP, means Querying from the knowledge graph.
For that won't we need Neo4j?

**Reply by romain-fontugne** [reply]
Yes understanding neo4j and IYP is also recommended. Also note that the IHR website will soon include helper functions to query IYP (see https://github.com/InternetHealthReport/ihr-website/pull/719)

**Comment by 0xAnon0602** [comment]
`displays these dependencies with very simple line charts that highlights changes over time.`
Where do you display this on website ?

**Reply by m-appel** [reply]
This refers to the "AS Dependency" graphs on the main IHR website. See [here](https://ihr.iijlab.net/ihr/en-us/networks/AS2501?af=4&last=3&date=2024-02-23&rov_tb=routes&hege_tb=dependency&hege_dt=2024-02-23%2000%3A15) for an example of the University of Tokyo.

**Reply by 0xAnon0602** [reply]
So we can get AS Dependency somewhat like this as mentioned in the idea right?

Example of ASN 2501 (University of Tokyo)
![bca1ca7d-3417-438a-991f-4cbb329e392d](https://github.com/InternetHealthReport/gsoc/assets/71931823/e8fe4d65-955d-404f-a0d0-60192bacaa06)

**Reply by romain-fontugne** [reply]
yes, that's the idea. See also the connectivity graphs on [bgp.tools](https://bgp.tools/as/2501#connectivity) or [bgp.he.net](https://bgp.he.net/AS2501#_graph4). I think if we add prefixes we should group prefixes that have the same dependencies, otherwise it will get quickly too cluttered

**Reply by ishqDehlvi** [reply]
Hello @romain-fontugne I want to contribute on this topic. I have the knowledge of the IHR main vue website codebase. Is there any pre-requisite too to get started?

**Reply by 0xAnon0602** [reply]
@romain-fontugne 
I don't think we really need to show prefixes to show the AS dependency 
Perhaps we could employ [Force Directed Graphs](https://observablehq.com/@d3/force-directed-graph-component) to more effectively showcase the dependencies of an AS.
Additionally, this method could allow us to further depict the dependencies of dependent ASes in a more optimized manner. 

What are your thoughts on this approach?

**Reply by romain-fontugne** [reply]
Adding prefixes would really make it stand out compare to the other websites mentioned above. But maybe it is a bit too  complex for the first version. Force Graphs are ok, but something a bit more ordered would be better. We could have the monitored AS on the left then the directly connected ASes (dependencies) and those that are two hops aways, etc...
Now I realize that the graph above is not showing the correct edges. We have two data sources. 1) the AS dependencies from IHR API, and 2) the connectivity between graphs that we can get from IYP.

**Reply by 0xAnon0602** [reply]
For the prefix, I'm currently trying to figure out how to group them together so they don't clutter the graphs.

I was also confused about choosing the data source (IHR API or Neo4j IYP). I'll try to create a basic implementation of the graph as you suggested using the Neo4j IYP data. 
Let me know if the IYP data source is right for this, or if I should shift to the IHR API for it.

**Reply by 0xAnon0602** [reply]
@romain-fontugne 
I'm trying to determine the routes of ASes to reach the specific AS similar to the website you mentioned above .
Can you guide me on how to do this or where I can study more about it?

**Reply by romain-fontugne** [reply]
Sorry I missed that message, you can get that from IYP: itp.iijlab.net. The second query listed [here](https://blog.apnic.net/2023/09/06/understanding-the-japanese-internet-with-the-internet-yellow-pages/) does something a bit similar.

**Comment by AyeshaLoladia** [comment]
Hii @romain-fontugne . I'm interested in this project also I have a thorough knowledge about BGP and cloud computing but I'm facing difficulty navigating through the IHR website and using it efficiently. I read the contributor guidance book thoroughly but it has no information on navigating through the website. Can you please provide me with any resources that would help me 😊😊

**Reply by romain-fontugne** [reply]
Hi, I'm not sure I understand the question. Maybe the website documentation (https://ihr.iijlab.net/ihr/en-us/documentation) or the readme for the source code (https://github.com/InternetHealthReport/ihr-website) may help?

**Reply by AyeshaLoladia** [reply]
Okay, I'll read it. Thanks for the help !

**Comment by Arunjagan12** [comment]
Sir, I have a few doubts regarding this project.
What are the functionalities you expect to be implemented in the component(dependency graph). What are all the behaviours or user interactions we should consider when designing this component?
What level of detail should the graph display? Is there any specific metrics we should focus on??

**Reply by romain-fontugne** [reply]
This is a visualization, so there isn't too much functionalities that are required regarding user interactions. Mainly showing more details if the user select or hover certain parts of the graph.
The graph should be intuitive and detailed information can be shown only if the user select/hover some parts of the graph. The main metric we are interested in here is AS hegemony (a.k.a AS dependency)

**Comment by arsh331** [comment]
Hi @romain-fontugne, I came across this project and I feel my experience in Neo4J and previous D3.js network graph project (https://listenbrainz.org/explore/music-neighborhood/) can be a good fit for this project.

I have gone over the project details, website and the internet yellow pages repository and am thinking of making a VueJs prototype as a proposal. Please let me know if it is a good path to take for this project.

Thanks

**Reply by romain-fontugne** [reply]
Hi, 
Expertise in computer networking is also needed for querying IYP and designing the visualization. So it may be more efficient to discuss before implementing a prototype. It depends on confident you are with your initial design.

**Reply by arsh331** [reply]
Sure, that's a good idea. I will jot down a few things as I understand them. Please review them and let me know if I am on the right track.
1. I saw a demo for the topology graph for Tokyo University, [link](https://ihr.iijlab.net/ihr/en-us/networks/AS2501?af=4&last=3&date=2024-02-20&rov_tb=routes&hege_tb=bgpPlay&hege_dt=2024-02-23%2000%3A15). As of now, I believe it is an external element rendered on the page but I am assuming the we are looking for something along the same lines.
2. I could not find any network graph for the plotly library (Please correct me if I am wrong here) but found a VueJS component for the D3.JS network graph, [link](https://github.com/emiliorizzo/vue-d3-network). So, we can modify and use it for the project.
3. Right now I am thinking of making a prototype by getting some data from the IYP database and then rendering it as a graph.

One more thing which I am thinking which might not be super relevant now is that as far as I have read the documentation IYP is not yet integrated into the main IHR website yet. So, is adding endpoints for getting the data from the IYP database a part of the project scope as well.

Moreover, I did go over the contributor handbook [link](https://github.com/InternetHealthReport/gsoc/blob/main/ihr-contributor-handbook.md) and as you mentioned I would need understanding of computer networks for querying the IYP. Could you please point to any additional resources I can take a look at?

Thank you.

**Reply by arsh331** [reply]
Just to add to the message, I found this file in the codebase, [link](https://github.com/InternetHealthReport/ihr-website/blob/iyp-integration/src/plugins/IypApi.js). Is this the recommended way of interacting with the database and getting the data for the graph.

**Reply by 0xAnon0602** [reply]
The IYP integration to the IHR website is already done (documentation is need to be updated)
Yes, that's the file you need to interact with IYP.

**Reply by romain-fontugne** [reply]
Yes the integration is done but not deployed on the main website.
You don't have to do it with plotly, and yes that's how we interact with iyp.

**Reply by arsh331** [reply]
Ok thanks. Could you please let me know what would make for a good proposal for this project?

**Reply by arsh331** [reply]
Hi @romain-fontugne, I have a prototype up and running, [here](https://codesandbox.io/p/devbox/happy-cherry-cnhpw2?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522cluas9fvp000a3x5xn3lzn842%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522cluas9fvo00023x5xyuueciiy%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522cluas9fvo00073x5x1bf4abct%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522cluas9fvo00093x5xu07j0gzn%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B44.11166529%252C55.88833471%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cluas9fvo00023x5xyuueciiy%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluas9fvo00013x5x9urvn1u1%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522cluas9fvo00023x5xyuueciiy%2522%252C%2522activeTabId%2522%253A%2522cluas9fvo00013x5x9urvn1u1%2522%257D%252C%2522cluas9fvo00093x5xu07j0gzn%2522%253A%257B%2522id%2522%253A%2522cluas9fvo00093x5xu07j0gzn%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluas9fvo00083x5xkqlzudqf%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522pnpm%2520run%2520dev%2522%252C%2522port%2522%253A5173%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522activeTabId%2522%253A%2522cluas9fvo00083x5xkqlzudqf%2522%257D%252C%2522cluas9fvo00073x5x1bf4abct%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cluas9fvo00033x5x46j8fulr%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522pnpm%2520run%2520dev%2522%257D%252C%257B%2522id%2522%253A%2522cluas9fvo00043x5xd1tni6hr%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clu7i6akm00a1dihu4zdr1quj%2522%257D%252C%257B%2522id%2522%253A%2522cluas9fvo00053x5xs4rm39pk%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clu7i95xk00a8dihu27np3bm5%2522%257D%252C%257B%2522id%2522%253A%2522cluas9fvo00063x5xg7zdhcex%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clu7iezd800andihua5ksd2y0%2522%257D%255D%252C%2522id%2522%253A%2522cluas9fvo00073x5x1bf4abct%2522%252C%2522activeTabId%2522%253A%2522cluas9fvo00033x5x46j8fulr%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A19.87179487179486%257D). I would really appreciate your feedback for the same.

**Reply by romain-fontugne** [reply]
See my comment above:
> Force Graphs are ok, but something a bit more ordered would be better. We could have the monitored AS on the left then the directly connected ASes (dependencies) and those that are two hops aways, etc...
> Now I realize that the graph above is not showing the correct edges. We have two data sources. 1) the AS dependencies from IHR API, and 2) the connectivity between graphs that we can get from IYP.

**Reply by arsh331** [reply]
So I tried to research some possible solutions to making the graph more ordered by incorporating a visual hierarchy based on the number of hops. I came across this article [link](https://www.thousandeyes.com/blog/visualizing-bgp-routing-better-network-operations), Moreover, I have found a library for d3 [link](https://github.com/erikbrinkman/d3-dag), which can help this. Is it something that matches your vision for the project?
Thank you.

**Reply by arsh331** [reply]
I modified this example query to find dependencies at max 3 hops away, 
```
MATCH (a:AS {asn:2501})-[h:DEPENDS_ON {af:4}]->(d:AS)
WITH a, COLLECT(DISTINCT d) AS dependencies
UNWIND dependencies as d
MATCH p = allShortestPaths((a)-[:PEERS_WITH*1..3]-(d))
WHERE a.asn <> d.asn AND all(r IN relationships(p) WHERE r.af = 4) AND all(n IN nodes(p) WHERE n IN dependencies)
RETURN p
```
Based on the result, I have annotated the results 
![Screenshot from 2024-03-29 19-02-07](https://github.com/InternetHealthReport/gsoc/assets/55950385/bc0fe00a-d1ca-4862-811c-14cc64ffb5ab)
Is this something you are talking about?

**Reply by romain-fontugne** [reply]
Yes that's correct.

**Reply by arsh331** [reply]
Hi @romain-fontugne, So I am working on my proposal for this project and this is the general idea of the graph I have come up with. It is basically a tree with multiple parents.
![ihr_graph](https://github.com/InternetHealthReport/gsoc/assets/55950385/cba403b6-8d09-4dfa-87e2-2b737d43e63f)
Your feedback is much appreciated.

**Reply by romain-fontugne** [reply]
That's interesting. Does it means that you ignore the number of hops for AS dependencies?
This could give interesting visualizations for large networks like CDNs. But it might be overkill for small networks. I think it would be better if we could switch between origin AS and originated prefixes

**Reply by arsh331** [reply]
Yeah that's the part I am confused with, because the project outline mentions all three (origin, dependencies and prefixes) to be in the graph. I might be wrong, but I feel depicting both the dependency hops and originating prefixes will be hard in a single graph. On this note, it just occurred to me, that we can represent the below part of the AS as a tree with the hops and relationships b/w the AS dependencies.
![image](https://github.com/InternetHealthReport/gsoc/assets/55950385/dd035231-7ce3-4859-a2c5-b65865388110)

For the next part we can use the above shown prefixes graph. 
![image](https://github.com/InternetHealthReport/gsoc/assets/55950385/25a87636-7581-47bb-b019-66d02ee4ffde)


Does it look like a solution to you. And I am really sorry for being late in the submission process.

**Reply by romain-fontugne** [reply]
You can see the origin AS as a way to represent all originated prefixes. The originated prefixes are within the origin AS. I understand it is confusing, but from the picture you sent earlier the left part (Origin AS and dependencies) shows the overall dependencies of the AS and the right part (Originated Prefixes and dependencies) shows the detailed dependencies. So yes it is hard to show both at the same time. One way would be to have a button that would switch between the two (overview/detailed) views

**Reply by arsh331** [reply]
That sounds like a good plan. I think I can work with this idea and can create a proposal with the same. I will try to complete it as soon as possible and will update you. Thank you

**Reply by 0xAnon0602** [reply]
Won't displaying the detailed view of all prefixes soon clutter the graph, making it nearly unreadable due to the large number of prefixes connected to a single AS?

> You can see the origin AS as a way to represent all originated prefixes. The originated prefixes are within the origin AS. I understand it is confusing, but from the picture you sent earlier the left part (Origin AS and dependencies) shows the overall dependencies of the AS and the right part (Originated Prefixes and dependencies) shows the detailed dependencies. So yes it is hard to show both at the same time. One way would be to have a button that would switch between the two (overview/detailed) views

**Reply by romain-fontugne** [reply]
yes, this will definitely happen. Especially for large networks. The best would be group all prefixes that have the exact same AS dependency values, and then the graph would show groups of prefixes
